src/senaite/core/z3cform/widgets/datetimewidget.py

```diff
@@ -218,27 +217,60 @@ def get_time(self, value):
             return u""
         return dtime.date_to_string(dt, HOUR_FORMAT)
 
-    def date_now(self, offset=0):
-        """Get the current date without time component
-        """
-        ts = datetime.now().strftime(DATE_FORMAT)
-        dt = datetime.strptime(ts, DATE_FORMAT)
-        if offset:
-            dt = dt + timedelta(offset)
-        return dt
-
     def attrs(self):
         """Return the template attributes for the date field
 
         :returns: dictionary of HTML attributes
         """
-        attrs = {}
-        today = self.date_now().strftime(DATE_FORMAT)
-        if self.datepicker_nofuture:
-            attrs["max"] = today
-        if self.datepicker_nopast:
-            attrs["min"] = today
-        return attrs
+        context = self.get_context()
+        min_date = self.get_min_date(context)
+        max_date = self.get_max_date(context)
+        return {
+            "min": dtime.date_to_string(min_date, DATE_FORMAT),
+            "max": dtime.date_to_string(max_date, DATE_FORMAT),
+        }
+
+    def get_min_date(self, instance):
+        """Returns the minimum datetime supported by this widget and instance
+        """
+        return self.resolve_date(self.min, instance, datetime.min)
+
+    def get_max_date(self, instance):
+        """Returns the maximum datetime supported for this widget and instance
+        """
+        return self.resolve_date(self.max, instance, datetime.max)
+
+    def resolve_date(self, thing, instance, default):
+        """Resolves the thing passed in to a DateTime object or None
+        """
+        if not thing:
+            return default
+
+        date = api.to_date(thing)
+        if api.is_date(date):
+            return date
+
+        if api.is_string(thing) and thing in ["current", "now"]:
+            return datetime.now()
+
+        if callable(thing):
+            value = thing()
+            return self.resolve_date(value, instance, default)
+
+        if hasattr(instance, thing):
+            value = getattr(instance, thing)
+            return self.resolve_date(value, instance, default)
+
+        if api.is_string(thing):
+            obj = api.get_object(instance, None)
+            fields = obj and api.get_fields(instance) or {}
+            field = fields.get(thing)
+            if field:
+                value = field.get(instance)
+                return self.resolve_date(value, instance, default)
```

Comment:
I don't like that the widget has all of this knowledge about the instance, fields and date retrieval baked in.

Why don't we make the `min` and `max` properties callables instead of simply strings?
Then we could look up the value directly in the field and store the calculated value.

I think we did the same approach in the reference fields to look up the query or column definition.